### ISSUE TITLE
Well-described metrics + dedicated documentation

### ISSUE THREAD

**Issue by ebenzecri** [scylladb/scylla#8888]
**Describe the feature and the current behavior/state.**
Metrics provided are poorly described. Here you have some examples:
`scylla_database_counter_cell_lock_acquisition: The number of acquired counter cell locks.` `scylla_database_counter_cell_lock_pending: The number of counter updates waiting for a lock.`

There's no documentation about them, so it's pretty impossible to know exactly what a metric is if there's no extra information about it. A one-liner description could help, but Scylla should have dedicated documentation to better explain each one.

**Who will benefit with this feature?**
Any person that works as Ops, Support, Architect, etc which uses metrics to get conclusions (and maybe taking decisions based on them).

**Comment by nyh** [comment]
I agree. A one-line title is better than nothing but very often when I don't understand what a certain metric really measures, this one line does not help me understand better :-( Indeed, we need much more detailed and unambiguous descriptions of each metric - descriptions that don't assume that the user is already an expert on these metrics or Scylla internals.

For consistency I think these long descriptions should be in the code, either as comments or as strings that don't get shown with the default metric HTTP request (but perhaps shown by some alternative HTTP request?), but we should be able to extract them into documentation on the site and/or `docs/`, **and** to make them somehow visible in the Grafana UI when you want to learn more about some metric.

P.S. I wrote the document https://github.com/scylladb/scylla/blob/master/docs/design-notes/metrics.md which is all about our metrics, **except** it doesn't list the metrics.. It suggests you should look at http://scyllanode:9180/metrics for a list of metrics and their explanations. However, @ebenzecri is right - many metric descriptions in that output aren't clear enough for non-experts.

**Comment by nyh** [comment]
We just realized in #10258 that we don't really know what `scylla_cache_row_hits` measures (a user saw it is almost always zero, and we don't understand why), or what `scylla_cache_reads` really measures (how is a "cache read" different from just a read - in what cases would a read not be counted as a cache read, or are there no such cases?).

Beyond the problem that we don't have good-enough documentation for the metrics, we also don't have good enough regression tests for them: If the aforementioned metrics are supposed to be incremented in certain cases, and not be incremented in other cases, we should have a test that verifies this - so that if I suddenly notice as we did in #10258 that the metric `scylla_cache_rows_hits` is zero, I could at least trust that this metric is correct, and didn't break a year ago and nobody noticed.

**Comment by nyh** [comment]
In duplicate issue #9249 there is some text on the different LWT metrics. I don't understand where that text came from (it mentions some "draft") but it is probably worth building on.

**Comment by nyh** [comment]
In the ScyllaDB Users mailing list thread https://groups.google.com/g/scylladb-users/c/eYbXsP0F8vw/m/H1POqB6HBgAJ, a user asked what does the `scylla_storage_proxy_coordinator_write_latency_count` metric count. His question might indicate that this counter has some sort of bug, but even before that - it isn't documented anywhere what this counter is supposed to be counting. `storage_proxy.cc` has this 5-word explanation:

> The general write latency histogram

which I didn't find particularly helpful (what does "general" refer here too? does it count coordinator or replica writes? does it count all writes? if something is a histogram, is the "count" for it accurate or approximate?)

**Comment by tomer-sandler** [comment]
@DoronArazii could you maybe see if @amnonh and/or @annastuchlik has the BW to start working on refactoring our metrics description to be better described with proper documentation? Thanks!
cc @tzach

**Comment by tzach** [comment]
Metrics descriptions should be fixed individually, each by an engineer who understands the relevant code.
Creating a task to fix *all* metrics is useless.

**Comment by nyh** [comment]
@tzach I'm not disputing that the people who should write these descriptions need to be engineers (in plural) which understand the code. However, this issue is **very much** needed because right now even if an engineer had the urge to explain one metric (as you suggested), he has no where to put this explanation! Our code sadly forces developers to write a single-line (often just 5 word) description of each metric, so that's what they do. Even if a developer wants to to write a paragraph-long or page-long explanation of some metric, he or she has nowhere to put it. Then, reviewers can't ask developers who add new metrics to write a lengthier description anywhere, because there is no such place.  There should be such a place, and this issue is about it.

Of course, we need to discuss where this place should be. Some people claim it should be **in the code** because separate documentation will become out of data and because in-code documentation can perhaps be passed to other code (e.g., perhaps we can add lengthy explanations to Grafana?). Other people suggest it should be **in documentation**, which became more realistic now that the documentation is part of the source repository so code patches can include documentation changes together in the same patch. In any case, whatever approach we choose, I think that what we have no - vague 5-word descriptions for each metric - is simply not good enough for people who don't use metrics often so when they need to do so, they don't remember what they mean.

**Comment by amnonh** [comment]
> single-line (often just 5 word) description of each metric, so that's what they do. Even if a developer wants to to write a paragraph-long or page-long explanation of some metric, he or she has nowhere to put it. Then, reviewers can't ask developers who add new metrics to write a lengthier description anywhere, because there is no such place. There should be such a place, and this issue is about it.
> 

We have the description part in the **code** we already tell Prometheus not to scrap the help lines, so when you reach from a browser you get the metrics with the help line and when Promethtue reads it, it does not get it.

**Comment by nyh** [comment]
> We have the description part in the **code** we already tell Prometheus not to scrap the help lines, so when you reach from a browser you get the metrics with the help line and when Promethtue reads it, it does not get it.

I don't understand this answer. As you know very well, currently all "descriptions" are vague single lines - and often much less than a single line (such as the super-helpful line _"The general write latency histogram"_ I mentioned above). Do you expect developers to expand these super-short descriptions to longer descripts? If so, how long should or can they be? Can they be a whole paragraph? A whole page? Will such long descriptions not hurt Prometheus performance? Will they be visible in a useful form in Grafana?

Again, I think the fact is that the metrics are **not** described well enough, and if you don't already know what a metric counts, the "description" we have for it often doesn't help at all.

My personal opinion is that (sorry Knuth...) documentation should be *with* the code, but not *in* the code. We need a long description for metrics, we need an ability to visually group several metrics into one section in the document (so we don't need to repeat common explanations that are relevant to 5 different metrics), we need to support links (e.g., the counter-related metric which started this issue can contain a link to the counter documentation, to make it clear which "counter" this relates to, and perhaps that document also explains why "locks" are used by these counters). All of this is difficult or unnatural to do inside the C++ code. BUT, if the decision will be to do this inside the code, I can live with that too. It will be better than the existing situation.

**Comment by ricardoborenstein** [comment]
I've seen many comments expressing frustration with the lack of scylla-monitoring documentation. Improving it would greatly enhance the user experience and onboarding of new employees. 
We can't continue ignoring this issue.
We need to provide explanations for each graph in every dashboard.
and it needs to start to be prioritized. 

A lot of tickets and hours of work would be avoided if the users would be able to understand the graphs, at this point, we should only have an explanation for metrics mentioned on the graphs.

@harel-z @mykaul  - How can we prioritize it?

**Comment by tomer-sandler** [comment]
> I've seen many comments expressing frustration with the lack of scylla-monitoring documentation. Improving it would greatly enhance the user experience and onboarding of new employees. We can't continue ignoring this issue. We need to provide explanations for each graph in every dashboard. and it needs to start to be prioritized.
> 
> A lot of tickets and hours of work would be avoided if the users would be able to understand the graphs, at this point, we should only have an explanation for metrics mentioned on the graphs.
> 
> @harel-z @mykaul - How can we prioritize it? @tomer-sandler @gcarmin - FYI

+1000

**Comment by mykaul** [comment]
It'll be more constructive if we can get several illusive ones and see how we solve those and learn from it rather than 'let's fix them all and now' attitude. Especially since metrics are not well 'owned' and are definitely 'distributed' when it comes to engineering responsibility, it is a bit challenging.  

@ricardoborenstein - if you can provide me with a select few that are mysterious enough that we need to handle them in higher priority, I'll work and get them improved, then we'll see how we can scale that process.

**Comment by gcarmin** [comment]
@mykaul  - I would argue that what we are looking to have is a "Scylla monitoring for dummies" kind of doc.
every time we have a new member in the service organisation the one thing that takes the most time to learn and understand is the monitoring stack.....
It can start as internal notion page that we'll move parts of it to public docs.
again this is my POV.

**Comment by ricardoborenstein** [comment]
@mykaul  My apologies, I did not mean to come across as forceful. I'm just trying to gain momentum on an issue that has gone on for 2 years and significantly impacts our users as well as ourselves.

I agree that tackling a few of the trickier issues together would be a more constructive approach than attempting to fix everything at once, especially since metrics ownership and responsibility seem to be distributed across engineering.

Perhaps we could start by filling in the missing information from the existing dashboards. Once we make progress on those, we'll have a better idea of how to scale the process.

Here is one example from the detailed dashboard, which has 4 graphs lacking information.
<img width="1997" alt="image" src="https://user-images.githubusercontent.com/78426391/224942501-d7f579d8-bf00-4242-9a0b-f2f8cec058d2.png">

**Comment by mykaul** [comment]
I see those descriptions missing from https://github.com/scylladb/scylla-monitoring/blob/8e5f5bcd87faaad3b345de68045493790817da6f/grafana/scylla-detailed.template.json#L194

@amnonh - who should be adding them?
(also, probably worth co-locating the description and title entries next to each other for easier browsing)

**Comment by amnonh** [comment]
I'm adding descriptions to the panels, the question is what the description
is, in this case, it's probably
reads/writes request on the replica node (which is equivalent to the title
- that's why it's not there).

**Comment by ricardoborenstein** [comment]
Perhaps we could add an explanation of how the balance is calculated and what it means if unbalanced.

**Comment by mykaul** [comment]
> > It'll be more constructive if we can get several illusive ones and see how we solve those and learn from it rather than 'let's fix them all and now' attitude.
> 
> I don't think anyone has this "fix them all and now" attitude. My attitude, for example, is what we must fix now is to add 1. a **way** to add documentation for individual metrics and groups of metrics, and 2. a way to show (or point with a link) to this documentation from Grafana.

I was under the impression the description is right next to the title in the JSON templates of the dashboards, for Grafana.

> 
> Right now, even a well-meaning developer which wants to write a whole book (or just one paragraph) about a specific metric has nowhere to write it.

Should we add it here - https://monitoring.docs.scylladb.com/stable/reference/index.html ?

> 
> > @ricardoborenstein - if you can provide me with a select few that are mysterious enough that we need to handle them in higher priority, I'll work and get them improved, then we'll see how we can scale that process.
> 
> In my experience **all** metrics are mysterious, and **all** deserve a long (at least a paragraph, often more) description. Of course we don't need to add them all at once and there are some which are more important, more mysterious - or both - than the others and should be explained first. But the mechanisms to show these new longer descriptions have to come before the actual descriptions.

**Comment by nyh** [comment]
> > I don't think anyone has this "fix them all and now" attitude. My attitude, for example, is what we must fix now is to add 1. a **way** to add documentation for individual metrics and groups of metrics, and 2. a way to show (or point with a link) to this documentation from Grafana.
> 
> I was under the impression the description is right next to the title in the JSON templates of the dashboards, for Grafana.

To be honest, I wasn't even aware that had these descriptions in the scylla monitoring repo! However, I am still not happy about this solution, for the following reasons:

1. When a developer adds a metric, they write a 5-word description for it in scylla.git. The longer description we have now is in a completely different repository (scylla-monitoring.git), and the scylla.git reviewers don't (and can't) insist that the developer *also* adds a longer description. Most don't. Moreover, nothing keeps the two descriptions tied.
2. The long description is tied to a specific panel. If the same metric appears in two panels, its long description can diverge.
3. The "long" description may not be long enough. I looked at a random panel (Alternator), most metrics had no description, and the first one that had, the description was "Number of nodes that reported their status as Starting or Joining". How would a beginner know what's the difference between those states are or why it's interesting? Why is "Joining" more interesting than "Leaving" or "Moving"? Ideally I would have liked to see a link to a document which explains this metric in even more, including links to a document about the different states and their meanings 

> 
> > Right now, even a well-meaning developer which wants to write a whole book (or just one paragraph) about a specific metric has nowhere to write it.
> 
> Should we add it here - https://monitoring.docs.scylladb.com/stable/reference/index.html ?

I think it should be in scylla.git - each metric defined just in one place (not short version in one place, long version in a second place, even longer version in third place). From there we could have a script that converts it to other formats (C++, JSON, .md, or whatever).

**Comment by amnonh** [comment]
We should be carefull not to mix metrics and dashboard panels (i.e. graphs).
In the famous data-information-knowldge pyramid, 
metrics are data, they are Scyllas version specific and should be tied closely to scylla repository.

Most of the time, graphs are information, for example taking the derivative of the total number of request to get the requests per second.

Beside trivial cases like latency or number of reads/writes. It helps to add an explenation to the panels, that is part of the monitoring and that is monitoring version specific, so it should be part of the monitoring repository.

Note, that in most situations there's no iff match between metrics and knowlge.
Take for example requests per shard.
A pick or drop in a single shard indicates that something may be wrong, but you'll need more information to know what that something is.
Could be a driver, a hot partition, maybe some background process fight over the core.

**Comment by ricardoborenstein** [comment]
We can describe what the missing dashboard information means and how users can look into this further.
We need to start listing the missing dashboard's information and then add descriptions.
My goal is to turn this discussion into action, and for that, we need a plan.
The plan would be:
List the missing dashboard information.
Ask experienced people to describe what each graph means for the end customer. (I think the solutions team should do this.)
Next, with the list and descriptions in hand, I can review the list with the solutions team and improve as needed.

**Comment by mykaul** [comment]
@ricardoborenstein - sounds like a good plan. I believe the team can directly contribute descriptions into scylla-monitoring and it'll be reviewed.

**Comment by ricardoborenstein** [comment]
I'll list the missing items and report back here.

**Comment by annastuchlik** [comment]
@nyh I totally agree with you that each metric should be described in one place. I'm just not sure why you think it should be in scylla.git. What would be the reason not to add it to the Reference section in the Monitoring docs? https://monitoring.docs.scylladb.com/stable/reference/index.html
I could create a metrics subfolder in https://github.com/scylladb/scylla-monitoring/tree/master/docs/source/reference where each description could be added in a separate file.

**Comment by amnonh** [comment]
It's easy enough to creat a table for the metrics description so when someone is going to add a metric, they would need to add an entry to that table.

It will require some changes on the Promethues server side. It's still differ from panel description/documentation

**Comment by tomer-sandler** [comment]
I don't have enough fingers in my body to count the amount of users and customers that keep telling us (for years now) that the only thing they can deduct from our monitoring is the Ops/request/throughput + Latency + Storage utilization data points.
Almost EVERY panel and EVERY metric is a mystery. Not to mention, how to use it (or a combination of metrics/panels), to deduct actionable intel from them. Same goes to our Advisor and Monitoring warnings...

**Comment by mykaul** [comment]
(Another target for more elaborate description of metrics is https://docs.datadoghq.com/integrations/scylla/ )

**Comment by tomer-sandler** [comment]
> (Another target for more elaborate description of metrics is https://docs.datadoghq.com/integrations/scylla/ )

@mykaul are you suggesting we send our customers to this DD page to learn more about Scylla Metrics?

**Comment by mykaul** [comment]
> 

No, just pointing out that we need to remember to update yet another location.

**Comment by tomer-sandler** [comment]
> I'll list the missing items and report back here.

Hello @ricardoborenstein , were you able to make progress on this?

**Comment by ricardoborenstein** [comment]
Yes,  I just shared the file with you.

**Comment by mykaul** [comment]
> Yes, I just shared the file with you.

Thanks - very useful. Two action items that I have in mind:
1. We can copy most/some of the descriptions from Scylla core. I've taken a quick look and added one as an example.
2. We should possibly add a Github action that ensures every JSON element with a 'title' sub-element should have a 'description' one as well, so we won't be adding more such description-less items.

**Comment by annastuchlik** [comment]
@ricardoborenstein Could you share the file with me, too?

**Comment by ricardoborenstein** [comment]
Hi Team, 

any updates here?
I don't see that the Excel file was updated a while ago and I don't want to see this fading away

**Comment by mykaul** [comment]
Let me take a stab at it, but it will take some time.

**Comment by annastuchlik** [comment]
To document the metrics, I'd need a .yaml file (or another format, tbd) that could be included in the documentation for each release.

**Comment by mykaul** [comment]
> To document the metrics, I'd need a .yaml file (or another format, tbd) that could be included in the documentation for each release.

When I'll find some time, I'll fish the descriptions from core, put them into the spreadsheet. From there, I hope to send a PR to fix the YAML - for items that simply lack the 'description' entity. So no need for anything else (but time).

**Comment by tzach** [comment]
> When I'll find some time, I'll fish the descriptions from core, put them into the spreadsheet. From there, I hope to send a PR to fix the YAML - for items that simply lack the 'description' entity. So no need for anything else (but time).

Its a lot fo work, that break the first time the code is updated.

There are two ways to automate this:

```mermaid
graph TD;
    Code --> YAML
    YAML --> Docs
```

```mermaid
graph TD;
    YAML --> Code
    YAML --> Docs
```

YAML here can be any usful represenation. The important part, all transation are automated into Make/Github actions/etc

**Comment by mykaul** [comment]
> > When I'll find some time, I'll fish the descriptions from core, put them into the spreadsheet. From there, I hope to send a PR to fix the YAML - for items that simply lack the 'description' entity. So no need for anything else (but time).
> 
> Its a lot fo work, that break the first time the code is updated.


Agreed, but it's also cross-projects, which makes it slightly more challenging. I rather start and fill in just the missing pieces - those that I /can/ find in Scylla core.

**Comment by tzach** [comment]
Why is it cross projects?
Both docs and code are in the same project.

Propogarte the info to Scylla Monitoring is harder.

**Comment by mykaul** [comment]
> Why is it cross projects? Both docs and code are in the same project.

The descriptions end up in Grafana, so belong to monitoring project.
I take them from the code in Core.

> 
> Propogarte the info to Scylla Monitoring is harder.

Yes, this is what I'm working on. Once we have the descriptions in the Monitoring YAML, I think it should be fairly easy to extract to docs format.

**Comment by mykaul** [comment]
Update: went over several (in te spreadsheet) , added some, noticed some already have a description (so unsure why they are in the spreadsheet?), some appear in multiple panels (yay for us - less work?). Only took the descriptions available from either Scylla core or Seastar, did not rewrite anything. Some ('Interface Rx Packets') perhaps do not need a description?

**Comment by amnonh** [comment]
As a first step, I created https://github.com/scylladb/scylla-monitoring/pull/1966
It takes all the descriptions from the current metrics, and add description for all the panels that had no description.
Not included yet, Alternator, Manager and node_exporter (os)

**Comment by annastuchlik** [comment]
@mykaul We've added automation to document the metrics, and they are all listed here:
https://docs.scylladb.com/manual/stable/reference/metrics.html

Can we close this issue?

**Comment by nyh** [comment]
> [@mykaul](https://github.com/mykaul) We've added automation to document the metrics, and they are all listed here: https://docs.scylladb.com/manual/stable/reference/metrics.html
> 
> Can we close this issue?

The original issue description said:

**A one-liner description could help, but Scylla should have dedicated documentation to better explain each one.**

The new page https://docs.scylladb.com/manual/stable/reference/metrics.html includes just these "one liner descriptions" and nothing more. So the new page is better than nothing, but I don't know if it fits the "spirit" of the original request in the issue.

I gave in one of the comments above (several years ago) the example of `scyllacache_row_hits` that puzzled me personally (in https://github.com/scylladb/scylladb/issues/10258). Maybe it's not the best example, I'm sure we can find even better examples with even more puzzling metrics. This metric still has a half-line description in metrics.html.

**Comment by annastuchlik** [comment]
@nyh I guess you're right. We have something on a page now that is helpful, but the user may need more information than just a brief description. I'll leave the issue open and assigned to @mykaul.

There's nothing I can do on my end now. We can publish any kind of information in the docs when given an input. A spreadsheet or an external source won't be a good way out, because:
- Each version has a metric update, so we need to ensure the descriptions are correct per version.
- The descriptions must be easy to update by the engineers who add/update them.

This means we need to pull the description of the metrics directly from the repository - it can be any folder or file, but it must live together with the code.
